File: templates/stack/monotonic_stack.go
package stack

func NextGreaterElements(nums []int) []int {
	result := make([]int, len(nums))
	for i := range result {
		result[i] = -1
	}

	st := []int{}
	for i, v := range nums {
		for len(st) > 0 && nums[st[len(st)-1]] < v {
			idx := st[len(st)-1]
			st = st[:len(st)-1]
			result[idx] = v
		}
		st = append(st, i)
	}
	return result
}

func DailyTemperatures(temps []int) []int {
	result := make([]int, len(temps))
	st := []int{}

	for i, t := range temps {
		for len(st) > 0 && temps[st[len(st)-1]] < t {
			idx := st[len(st)-1]
			st = st[:len(st)-1]
			result[idx] = i - idx
		}
		st = append(st, i)
	}
	return result
}

func LargestRectangleArea(heights []int) int {
	best := 0
	st := []int{}
	extended := append(append([]int{}, heights...), 0)

	for i, h := range extended {
		for len(st) > 0 && extended[st[len(st)-1]] > h {
			height := extended[st[len(st)-1]]
			st = st[:len(st)-1]

			left := -1
			if len(st) > 0 {
				left = st[len(st)-1]
			}
			width := i - left - 1
			area := height * width
			if area > best {
				best = area
			}
		}
		st = append(st, i)
	}
	return best
}

File: templates/stack/monotonic_stack.py
def next_greater_elements(nums: list[int]) -> list[int]:
    result = [-1] * len(nums)
    stack: list[int] = []

    for i, value in enumerate(nums):
        while stack and nums[stack[-1]] < value:
            idx = stack.pop()
            result[idx] = value
        stack.append(i)

    return result


def daily_temperatures(temps: list[int]) -> list[int]:
    result = [0] * len(temps)
    stack: list[int] = []

    for i, temp in enumerate(temps):
        while stack and temps[stack[-1]] < temp:
            idx = stack.pop()
            result[idx] = i - idx
        stack.append(i)

    return result


def largest_rectangle_area(heights: list[int]) -> int:
    best = 0
    stack: list[int] = []
    extended = heights + [0]

    for i, height in enumerate(extended):
        while stack and extended[stack[-1]] > height:
            popped_height = extended[stack.pop()]
            left = stack[-1] if stack else -1
            width = i - left - 1
            best = max(best, popped_height * width)
        stack.append(i)

    return best
